### PR TITLE
Restore support for using V8 as the Wasm JIT interpreter

### PR DESCRIPTION
**Pull request by steven-johnson**
I've taken the liberty to branch https://github.com/halide/Halide/pull/6429 (with permission from @ngzhian) and finish up the work needed to land it. 

This passes locally. The V8-specific code won't (yet) get tested on the buildbots (the script and bots need attention to do so) but it can land without that.

### CODE REVIEW

**Comment by steven-johnson** [comment]
OSX-x86 bot is down for repairs, probably OK to ignore for this

**Review by alexreinking — CHANGES_REQUESTED** [review]

File: dependencies/wasm/CMakeLists.txt [hunk 1]
```diff
@@ -48,6 +53,25 @@ if (WITH_WABT)
...
+    if (NOT V8_INCLUDE_PATH)
+        message(FATAL_ERROR "Please set V8_INCLUDE_PATH on the CMake command line.")
+    endif()
+    if (NOT V8_LIB_PATH)
+        message(FATAL_ERROR "Please set V8_LIB_PATH on the CMake command line.")
+    endif()
+
+    message(STATUS "Using V8")
+    add_library(Halide_v8 INTERFACE)
+    target_sources(Halide_v8 INTERFACE ${V8_LIB_PATH})
```
[the remark this answers is not in the document]
Comment (steven-johnson) [reply dependencies/wasm/CMakeLists.txt:67]: Good question, I just took this from the source PR branch as-is and it worked -- I will make any and all CMake changes as you suggest :-)

File: README_webassembly.md [hunk 2]
```diff
@@ -80,6 +80,55 @@ environment and are disabled).
 In sum: don't plan on using Halide JIT mode with Wasm unless you are working on
 the Halide library itself.
 
+## Using V8 as the interpreter
+
+There is experimental support for using V8 as the interpreter in JIT mode, rather than WABT.
+This is enabled by the CMake command line options `-DWITH_V8=ON -DWITH_WABT=OFF` (only one of them can be used at a time).
+You must build V8 locally V8, then specify the path to the library and headers as CMake options.
+This is currently only tested on x86-64-Linux and requires v8 version 9.8.147 as a minimum.
```
[the remark this answers is not in the document]
Comment (steven-johnson) [reply README_webassembly.md:88]: Hm, that could be made to work I guess.

File: README_webassembly.md [hunk 3]
```diff
@@ -80,6 +80,55 @@ environment and are disabled).
...
+
+```
+$ cd /path/to/halide
+$ export HL_TARGET=wasm-32-wasmrt-wasm_simd128
+$ export HL_JIT_TARGET=${HL_TARGET}
+$ cmake -G Ninja \
+      -DWITH_WABT=OFF \
+      -DWITH_V8=ON \
+      -DV8_INCLUDE_PATH=$HOME/v8/v8/include \
+      -DV8_LIB_PATH=$HOME/v8/v8/out.gn/x64.release.sample/obj/libv8_monolith.a \
```
[the remark this answers is not in the document]
Comment (steven-johnson) [reply README_webassembly.md:123]: I would be quite surprised if such a thing existed.

File: dependencies/wasm/CMakeLists.txt [hunk 4]
```diff
@@ -48,6 +53,25 @@ if (WITH_WABT)
...
+    target_include_directories(Halide_v8
+        SYSTEM
+        INTERFACE
+        ${V8_INCLUDE_PATH})
+    set_target_properties(Halide_v8 PROPERTIES EXPORT_NAME v8)
```
[the remark this answers is not in the document]
Comment (alexreinking) [reply dependencies/wasm/CMakeLists.txt:72]: Okay, that's an acceptable answer.

File: README_webassembly.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (alexreinking) [reply README_webassembly.md:123]: If we don't care about packaging, then it doesn't matter.

File: README_webassembly.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (alexreinking) [reply README_webassembly.md:88]: It was just a stray thought... maybe worth it if we add a third backend (if one even exists?).

**Comment by alexreinking** [comment]
Okay, I think I'd like to see the `target_sources` / `target_link_libraries` comments addressed. If we don't care about packaging this, then none of the other stuff matters.

File: src/CMakeLists.txt [hunk 5]
```diff
@@ -401,6 +401,12 @@ if (TARGET Halide_wabt)
     target_compile_definitions(Halide PRIVATE WITH_WABT)
 endif ()
 
+if (TARGET Halide_v8)
+    target_link_libraries(Halide PRIVATE Halide_v8)
+    target_link_libraries(Halide PRIVATE ${V8_LIB_PATH})
```
[the remark this answers is not in the document]
Comment (steven-johnson) [reply src/CMakeLists.txt:406]: ok, this is weird: if I *remove* the second line, I get *duplicate symbols* errors when linking. Leaving it like this works. Investigating.

File: dependencies/wasm/CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (alexreinking) [reply dependencies/wasm/CMakeLists.txt:67]: Actually, this whole thing should be:

```
add_library(Halide::V8 STATIC IMPORTED)
set_target_properties(Halide::V8 PROPERTIES IMPORTED_LOCATION "${V8_LIB_PATH}")
target_include_directories(Halide::V8 INTERFACE "${V8_INCLUDE_PATH}")
```

and then in `src/CMakeLists.txt` link only to `Halide::V8`.

File: src/CMakeLists.txt (hunk 5 above)
[the remark this answers is not in the document]
Comment (steven-johnson) [reply src/CMakeLists.txt:406]: It turns out that both LLVM and V8 define a toplevel symbol named `__jit_debug_descriptor`, for GDB integration... which produces link errors, and the previous wonky CMake rules managed to defeat this via, uh, I dunno. V8 needs to be built with `v8_enable_gdbjit = false` to make this work.

[the remark this answers is not in the document]
Comment (alexreinking) [reply src/CMakeLists.txt:406]: Was the library appearing twice on the link line?

File: README_webassembly.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (steven-johnson) [reply README_webassembly.md:88]: I'm inclined to leave this as-is for now. Will revisit later if needed.

File: src/CMakeLists.txt (hunk 5 above)
[the remark this answers is not in the document]
Comment (alexreinking) [reply src/CMakeLists.txt:406]: This is a biiiig wtf. We should ensure we're building V8 without gdbjit or at _least_ get to the bottom of this and add a comment explaining what is going on here and why it works (for now).

[the remark this answers is not in the document]
Comment (steven-johnson) [reply src/CMakeLists.txt:406]: Yes, rebuilding without gdbjit solves the link issue. Why it works with the current/previous CMake rules is a matter I don't care to spend time investigating.

File: dependencies/wasm/CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (steven-johnson) [reply dependencies/wasm/CMakeLists.txt:67]: OK, this is weird; I changed the deps/wasm/CMakeLists.txt like so:

    message(STATUS "Using V8")
    add_library(Halide_V8 STATIC IMPORTED)
    set_target_properties(Halide_V8 PROPERTIES IMPORTED_LOCATION "${V8_LIB_PATH}")
    target_include_directories(Halide_V8 INTERFACE "${V8_INCLUDE_PATH}")

and in src/CMakeLists.txt, like so:

```
if (TARGET Halide_V8)
    target_link_libraries(Halide PRIVATE Halide_V8)
    target_compile_definitions(Halide PRIVATE WITH_V8)
else()
    message(FATAL_ERROR "V8 not found")
endif ()
```

...and it always fails at the fatal error. Is there something about "IF TARGET" that is wonky with imported libs?

[the remark this answers is not in the document]
Comment (alexreinking) [reply dependencies/wasm/CMakeLists.txt:67]: Ugh, yes... need to make it `STATIC IMPORTED GLOBAL`.

[the remark this answers is not in the document]
Comment (steven-johnson) [reply dependencies/wasm/CMakeLists.txt:67]: aw yeah

**Comment by steven-johnson** [comment]
PTAL

**Review by alexreinking — APPROVED** [review]

**Comment by alexreinking** [comment]
Is this good to merge? Have you looked at the win64 failure?

**Comment by steven-johnson** [comment]
There's no way thhe windows failure is related, but I will sync to head and retry.

**Pull request merged**